Use IHtmlGenerator / TagBuilder in all necessary tag helpers

SecTex opened the pull request:
Fixes #5322

SecTex committed:
use text() instead of html()

SecTex committed:
Merge branch 'dev' of https://github.com/abpframework/abp into SecTex/html-encoding

SecTex committed:
Use TagBuilder to create HTML elements

maliming commented:
hi @SecTex 

Thanks for your contributions, we have many tag helper so there will be a lot of work. If you don't have much time I can handle it .

SecTex committed:
Use IHtmlGenerator / TagBuilder in necessary tag helpers

SecTex commented:
Hi @maliming

It's fine, I'll take the time to revise all tag helpers.

SecTex committed:
Use IHtmlGenerator / TagBuilder in necessary tag helpers

Notes:
- Would be nice to completely remove the use of `AbpTagHelperService.RenderHtml`

SecTex committed:
reverted some changes to not break class signatures

SecTex committed:
reverted some changes to not break class signatures

SecTex committed:
fixed encoding issue

SecTex committed:
fixed encoding issue when using data uris

SecTex committed:
fixed typo

SecTex committed:
fixed encoding issue

SecTex committed:
escape html by default

SecTex committed:
Merge branch 'dev' of https://github.com/abpframework/abp into SecTex/html-encoding

maliming commented:
Thanks @SecTex  very much!

maliming commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/toastr/abp-toastr.js:
```diff
@@ -12,6 +12,12 @@
+        }
+        if (options.escapeHtml === undefined) {
+            options.escapeHtml = true;
+        }
```
I think we shouldn't change `options`. The developer should be handle it.

Commit:
Use TagBuilderExtensions instead of RenderHtml.

maliming commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/jquery/jquery-extensions.js:
```diff
@@ -55,7 +55,7 @@
                 //restore text
                 if ($buttonInnerSpan.length && $button.data('buttonOriginalText')) {
-                    $buttonInnerSpan.html($button.data('buttonOriginalText'));
+                    $buttonInnerSpan.text($button.data('buttonOriginalText'));
```
We can add `data-busy-html` attribute. or use a bool `data-busy-is-html`.

maliming commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/datatables/datatables-extensions.js:
```diff
@@ -27,6 +27,10 @@
             }
         };
 
+        var appendEncoded = function ($element, unencoded) {
+            $element.append($('<div/>').text(unencoded).html());
+        };
+
```
```
var htmlEncode = function (html) {
	return $('<div/>').text(html).html();
};
```

maliming commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/datatables/datatables-extensions.js:
```diff
@@ -41,7 +45,7 @@
                 }
 
-                $a.append(fieldItem.text);
+                appendEncoded($a, fieldItem.text);
```
`$a.append(htmlEncode(fieldItem.text));`

maliming commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/datatables/datatables-extensions.js:
```diff
@@ -85,9 +89,9 @@
 
             if (field.text) {
-                $dropdownButton.append(field.text);
+                appendEncoded($dropdownButton, field.text);
```
`$dropdownButton.append(htmlEncode(fieldItem.text));`

maliming commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/datatables/datatables-extensions.js:
```diff
@@ -85,9 +89,9 @@
+                appendEncoded($dropdownButton, field.text);
             } else {
-                $dropdownButton.append(localize("DatatableActionDropdownDefaultText"));
+                appendEncoded($dropdownButton, localize("DatatableActionDropdownDefaultText"));
```
`$dropdownButton.append(htmlEncode(localize("DatatableActionDropdownDefaultText")));`

SecTex committed:
revert a725b49

SecTex committed:
added `busy-text-is-html` attribute

SecTex committed:
`appendEncoded` -> `htmlEncode`

SecTex committed:
fixed demo page

SecTex commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Bootstrap/TagHelpers/Carousel/AbpCarouselItemTagHelperService.cs:
```diff
@@ -42,8 +42,12 @@ private void AddToContext(TagHelperContext context, TagHelperOutput output)
-            output.Content.SetHtmlContent(imgTag);
+            var img = new TagBuilder("img");
+            img.AddCssClass("d-block w-100");
+            img.Attributes.Add("src", TagHelper.Src);
```
This may break existing carousels.
I fixed the demo page in 7a1820c.

SecTex committed:
Merge branch 'dev' of https://github.com/abpframework/abp into SecTex/html-encoding

# Conflicts:
#	framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/datatables/datatables-extensions.js

maliming commented on framework/src/Volo.Abp.AspNetCore.Mvc.UI.Theme.Shared/wwwroot/libs/abp/aspnetcore-mvc-ui-theme-shared/jquery/jquery-extensions.js:
```diff
@@ -35,7 +35,12 @@
                     $button.data('buttonOriginalText', $buttonInnerSpan.html());
-                    $buttonInnerSpan.html($button.attr('data-busy-text'));
+                    
+                    if ($button.data('busy-text-is-html')) {
```
Is this should be `data-busy-text-is-html`?

SecTex replied:
I think `busy-text-is-html` is correct here, since I use jQuerys `data()`-function:

```javascript
var element = $('<button>').attr('data-busy-text-is-html', true);
» undefined

element.data('busy-text-is-html');
» true

element.attr('data-busy-text-is-html')
» "true"
```

maliming replied:
You are right, Thanks.

maliming committed:
Run yarn & gulp in Bootstrap.Demo project.
